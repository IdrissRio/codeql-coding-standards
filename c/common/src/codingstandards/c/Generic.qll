import cpp
import codingstandards.cpp.Macro
import codingstandards.cpp.MatchingParenthesis

string genericRegexp() { result = "\\b_Generic\\s*\\(\\s*(.+),.*" }

bindingset[input]
string deparenthesize(string input) {
  input = "(" + result + ")" and
  result = input.substring(1, input.length() - 1)
}

class GenericMacro extends Macro {
  string ctrlExpr;

  GenericMacro() { ctrlExpr = getBody().regexpCapture(genericRegexp(), 1).trim() }

  string getAParameter() { result = this.(FunctionLikeMacro).getAParameter() }

  string getControllingExprString() {
    if exists(string s | s = deparenthesize(ctrlExpr))
    then result = deparenthesize(ctrlExpr).trim()
    else result = ctrlExpr
  }

  /**
   * Whether the controlling expression of the `_Generic` expr in this macro's controlling
   * expression refers to one of this macro's parameters.
   */
  predicate hasControllingExprFromMacroParameter() {
    getControllingExprString().matches(getAParameter())
  }
}

class GenericMacroString extends string {
  GenericMacroString() { this = any(Macro m).getBody() and this.matches("%_Generic%") }
}

import MatchingParenthesis<GenericMacroString>

class ParsedGenericMacro extends Macro {
  ParsedRoot macroBody;
  Parsed genericBody;
  string beforeGenericBody;
  string afterGenericBody;

  ParsedGenericMacro() {
    macroBody.getInputString() = this.getBody() and
    exists(ParsedText genericText |
      genericText.getText().matches("%_Generic%") and
      genericBody = genericText.getParent().getChild(genericText.getChildIdx() + 1) and
      genericBody.getRoot() = macroBody
    ) and
    beforeGenericBody =
      textFrom(macroBody.getStartToken(), genericBody.getStartToken().getPrevious()) and
    (
      if exists(genericBody.getEndToken().getNext())
      then afterGenericBody = textFrom(genericBody.getEndToken().getNext(), macroBody.getEndToken())
      else afterGenericBody = ""
    )
  }

  string getAParameter() { result = this.(FunctionLikeMacro).getAParameter() }

  int getAParsedGenericCommaSeparatorOffset() {
    exists(ParsedText text |
      text.getParent() = genericBody and
      result = text.getStartToken().getStartPos() + text.getText().indexOf(",")
    )
  }

  int getAParsedGenericColonSeparatorOffset() {
    exists(ParsedText text |
      text.getParent() = genericBody and
      result = text.getStartToken().getStartPos() + text.getText().indexOf(":")
    )
  }

  int getParsedGenericCommaSeparatorOffset(int i) {
    result = rank[i](int index | index = getAParsedGenericCommaSeparatorOffset())
  }

  bindingset[start, end]
  int getParsedGenericColon(int start, int end) {
    result =
      min(int offset |
        offset = getAParsedGenericColonSeparatorOffset() and
        offset >= start and
        offset <= end
      )
  }

  predicate hasParsedFullSelectionRange(int idx, int start, int end) {
    idx = 1 and
    start = genericBody.getStartToken().getEndPos() and
    end = getParsedGenericCommaSeparatorOffset(idx)
    or
    not exists(getParsedGenericCommaSeparatorOffset(idx)) and
    start = getParsedGenericCommaSeparatorOffset(idx - 1) and
    end = genericBody.getEndToken().getStartPos()
    or
    start = getParsedGenericCommaSeparatorOffset(idx - 1) and
    end = getParsedGenericCommaSeparatorOffset(idx)
  }

  string getSelectionString(int idx) {
    exists(int start, int rawStart, int end |
      hasParsedFullSelectionRange(idx, rawStart, end) and
      (
        if exists(getParsedGenericColon(rawStart, end))
        then start = getParsedGenericColon(rawStart, end)
        else start = rawStart
      ) and
      result = genericBody.getInputString().substring(start, end)
    )
  }

  string getControllingExprString() { result = getSelectionString(1).trim() }

  bindingset[str, word]
  private int countWordInString(string word, string str) {
    result =
      max(int occurrence |
        exists(str.regexpFind("\\b" + word + "\\b", occurrence, _)) or occurrence = -1
      |
        occurrence + 1
      )
  }

  int expansionsOutsideExpr(string parameter) {
    parameter = getAParameter() and
    result =
      countWordInString(parameter, beforeGenericBody) +
        countWordInString(parameter, afterGenericBody)
  }

  int expansionsInsideSelection(string parameter, int idx) {
    parameter = getAParameter() and
    result = countWordInString(parameter, getSelectionString(idx))
  }

  int expansionsInsideControllingExpr(string parameter) {
    result = expansionsInsideSelection(parameter, 1)
  }

  int expansionsInsideAssociation(string parameter, int idx) {
    not idx = 0 and
    result = expansionsInsideSelection(parameter, idx + 1)
  }
}
