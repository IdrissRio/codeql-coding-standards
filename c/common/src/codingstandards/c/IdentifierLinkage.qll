import cpp

newtype TIdentifierLinkage =
  TIdentifierLinkageExternal() or
  TIdentifierLinkageInternal() or
  TIdentifierLinkageNone()

/**
 * In C, identifiers have internal linkage, or external linkage, or no linkage (6.2.2.1).
 *
 * The linkage of an identifier is used to, among other things, determine the storage duration
 * and/or lifetime of that identifier. Storage durations and lifetimes are often used to define
 * rules in the various coding standards.
 */
class IdentifierLinkage extends TIdentifierLinkage {
  predicate isExternal() { this = TIdentifierLinkageExternal() }

  predicate isInternal() { this = TIdentifierLinkageInternal() }

  predicate isNone() { this = TIdentifierLinkageNone() }

  string toString() {
    this.isExternal() and result = "external linkage"
    or
    this.isInternal() and result = "internal linkage"
    or
    this.isNone() and result = "no linkage"
  }
}

/**
 * Determine the linkage of a variable: external, or static, or none.
 *
 * The linkage of a variable is determined by its scope and storage class. Note that other types of
 * identifiers (e.g. functions) may also have linkage, but that behavior is not covered in this
 * predicate.
 */
IdentifierLinkage linkageOfVariable(Variable v) {
  // 6.2.2.3, file scope identifiers marked static have internal linkage.
  v.isTopLevel() and v.isStatic() and result.isInternal()
  or
  // 6.2.2.4 describes generally non-static file scope identifiers, which have external linkage.
  v.isTopLevel() and not v.isStatic() and result.isExternal()
  or
  // Note: Not all identifiers have linkage, see 6.2.2.6
  not v.isTopLevel() and result.isNone()
}
