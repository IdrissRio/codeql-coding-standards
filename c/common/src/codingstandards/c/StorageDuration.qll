import cpp

class DeclarationWithStorageDuration extends Declaration { }

newtype TStorageDuration =
  StorageDurationStatic() or
  StorageDurationAutomatic() or
  StorageDurationThread() or
  StorageDurationAllocated()

class StorageDuration extends TStorageDuration {
  predicate isStatic() { this = StorageDurationStatic() }

  predicate isAutomatic() { this = StorageDurationAutomatic() }

  predicate isThread() { this = StorageDurationThread() }

  predicate isAllocated() { this = StorageDurationAllocated() }

  string toString() { result = getStorageTypeName() + " storage duration" }

  string getStorageTypeName() {
    isStatic() and result = "static"
    or
    isAutomatic() and result = "automatic"
    or
    isThread() and result = "thread"
    or
    isAllocated() and result = "allocated"
  }
}
