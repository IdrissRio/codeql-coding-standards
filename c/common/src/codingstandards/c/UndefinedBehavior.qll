import cpp
import codingstandards.cpp.types.Pointers
import codingstandards.cpp.UndefinedBehavior

/**
 * Library for modeling undefined behavior.
 */
abstract class CUndefinedBehavior extends UndefinedBehavior { }

/**
 * A function which has the signature - but not the name - of a main function.
 */
class C99MainFunction extends Function {
  C99MainFunction() {
    this.getNumberOfParameters() = 2 and
    this.getType().getUnderlyingType() instanceof IntType and
    this.getParameter(0).getType().getUnderlyingType() instanceof IntType and
    this.getParameter(1)
        .getType()
        .getUnderlyingType()
        .(UnspecifiedPointerOrArrayType)
        .getBaseType()
        .(UnspecifiedPointerOrArrayType)
        .getBaseType() instanceof CharType
    or
    this.getNumberOfParameters() = 0 and
    // Must be explicitly declared as `int main(void)`.
    this.getADeclarationEntry().hasVoidParamList() and
    this.getType().getUnderlyingType() instanceof IntType
  }
}

class CUndefinedMainDefinition extends CUndefinedBehavior, Function {
  CUndefinedMainDefinition() {
    // for testing purposes, we use the prefix ____codeql_coding_standards`
    (this.getName() = "main" or this.getName().indexOf("____codeql_coding_standards_main") = 0) and
    not this instanceof C99MainFunction
  }

  override string getReason() {
    result =
      "main function may trigger undefined behavior because it is not in one of the formats specified by the C standard."
  }
}
