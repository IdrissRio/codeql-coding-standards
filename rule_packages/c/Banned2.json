{
  "MISRA-C-2012": {
    "RULE-21-24": {
      "properties": {
        "obligation": "required"
      },
      "queries": [
        {
          "description": "The standard functions rand() and srand() will not give high quality random results in all implementations and are therefore banned.",
          "kind": "problem",
          "name": "The random number generator functions of <stdlib.h> shall not be used",
          "precision": "very-high",
          "severity": "warning",
          "short_name": "CallToBannedRandomFunction",
          "tags": [
            "security",
            "external/misra/c/2012/amendment3"
          ]
        }
      ],
      "title": "The random number generator functions of <stdlib.h> shall not be used"
    }
  }
}