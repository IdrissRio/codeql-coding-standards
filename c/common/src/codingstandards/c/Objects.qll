import cpp
import codingstandards.c.StorageDuration
import codingstandards.c.IdentifierLinkage
import semmle.code.cpp.valuenumbering.HashCons
import codingstandards.cpp.Clvalues

/**
 * A libary for handling "Objects" in C.
 *
 * Objects may be stored in registers or memory, they have an address, a type, a storage duration,
 * and a lifetime (which is different than storage duration). Objects which are structs or arrays
 * have subobjects, which share the storage duration and lifetime of the parent object.
 *
 * Note: lifetime analysis is not performed in this library, but is available in
 * the module `codingstandards.cpp.lifetimes.LifetimeProfile`. In the future, these libraries could
 * be merged for more complete analysis.
 *
 * To get objects in a project, use the `ObjectIdentity` class which finds the following types of
 * objects:
 * - global variables
 * - local variables
 * - literals
 * - malloc calls
 * - certain temporary object expressions
 *
 * And direct references to these objects can be found via the member predicate `getAnAccess()`.
 * However, much of a project's code will not refer to these objects directly, but rather, refer to
 * their subobjects. The class `ObjectIdentity` exposes several member predicates for finding when
 * these subobjects are used:
 * - `getASubobjectType()`
 * - `getASubobjectAccess()`
 * - `getASubobjectAddressExpr()`
 *
 * These methods do not use flow analysis, and will not return a conclusive list of accesses. To
 * get better results here, this library should be integrated with flow analysis or the library
 * `LifetimeProfile.qll`.
 *
 * Additionally, subobjects are currently not tracked individually. In the future subobjects could
 * be tracked as a root object and an access chain to refer to them. For now, however, finding *any*
 * subobject access is sufficient for many analyses.
 *
 * To get the storage duration, `ObjectIdentity` exposes the member predicate
 * `getStorageDuration()` with the following options:
 * - `obj.getStorageDuration().isAutomatic()`: Stack objects
 * - `obj.getStorageDuration().isStatic()`: Global objects
 * - `obj.getStorageDuration().isThread()`: Threadlocal objects
 * - `obj.getStorageDuration().isAllocated()`: Dynamic objects
 *
 * Note that lifetimes are not storage durations. The only lifetime tracking currently implemented
 * is `hasTemporaryLifetime()`, which is a subset of automatic storage duration objects, and may
 * be filtered out, or selected directly with `TemporaryObjectIdentity`.
 */
final class ObjectIdentity = ObjectIdentityBase;

/**
 * A base class for objects in C, along with the source location where the object can be identified
 * in the project code (thus, this class extends `Element`), which may be variable, or may be an
 * expression such as a literal or a malloc call.
 *
 * Extend this class to define a new type of object identity. To create a class which filters the
 * set of object identities, users of this library should extend the final subclass
 * `ObjectIdentity` instead.
 */
abstract class ObjectIdentityBase extends Element {
  /**
   * The type of this object.
   *
   * Note that for allocated objects, this is inferred from the sizeof() statement or the variable
   * it is assigned to.
   */
  abstract Type getType();

  /* The storage duration of this object: static, thread, automatic, or allocated. */
  abstract StorageDuration getStorageDuration();

  /**
   * Get the nested objects within this object (members, array element types).
   *
   * Note that if a struct has a pointer member, the pointer itself is a subobject, but the value
   * it points to is not. Therefore `struct { int* x; }` has a subobject of type `int*`, but not
   * `int`.
   */
  Type getASubObjectType() { result = getADirectSubobjectType*(getType()) }

  /**
   * Get expressions which trivially access this object. Does not perform flow analysis.
   *
   * For dynamically allocated objects, this is a dereference of the malloc call.
   */
  abstract Expr getAnAccess();

  /**
   * Get expressions which trivially access this object or a subobject of this object. Does not
   * perform flow analysis.
   *
   * For dynamically allocated objects, this is a dereference of the malloc call or direct access
   * of the result of dereferencing the malloc call.
   */
  Expr getASubobjectAccess() { result = getASubobjectAccessOf(getAnAccess()) }

  /**
   * Get expressions which trivially take the address of this object or a subobject of this object.
   * Does not perform flow analysis.
   */
  Expr getASubobjectAddressExpr() {
    exists(Expr subobject |
      subobject = getASubobjectAccess() and
      (
        // Holds for address-of expressions.
        result = any(AddressOfExpr e | e.getOperand() = subobject)
        or
        // Holds for array-to-pointer conversions, which evaluate to a usable subobject address.
        exists(ArrayToPointerConversion c | c.getExpr() = subobject) and
        // Note that `arr[x]` has an array-to-pointer conversion, and returns the `x`th item by
        // value, not the address of the `x`th item. Therefore, exclude `arr` if `arr` is part of
        // an expression `arr[x]`.
        not exists(ArrayExpr a | a.getArrayBase() = subobject) and
        result = subobject
      )
    )
  }

  /**
   * Holds if the object has temporary lifetime. This is not a storage duration, but only objects
   * with automatic storage duration have temporary lifetime.
   */
  abstract predicate hasTemporaryLifetime();
}

/**
 * Finds expressions `e.x` or `e[x]` for expression `e`, recursively. Does not resolve pointers.
 *
 * Note that this does not hold for `e->x` or `e[x]` where `e` is a pointer.
 */
private Expr getASubobjectAccessOf(Expr e) {
  result = e
  or
  result.(DotFieldAccess).getQualifier() = getASubobjectAccessOf(e)
  or
  result.(ArrayExpr).getArrayBase() = getASubobjectAccessOf(e) and
  not result.(ArrayExpr).getArrayBase().getUnspecifiedType() instanceof PointerType
}

/**
 * Find the object types that are embedded within the current type.
 *
 * For example, a block of memory with type `T[]` has subobjects of type `T`, and a struct with a
 * member of `T member;` has a subobject of type `T`.
 *
 * Note that subobjects may be pointers, but the value they point to is not a subobject. For
 * instance, `struct { int* x; }` has a subobject of type `int*`, but not `int`.
 */
Type getADirectSubobjectType(Type type) {
  result = type.stripTopLevelSpecifiers().(Struct).getAMember().getADeclarationEntry().getType()
  or
  result = type.stripTopLevelSpecifiers().(ArrayType).getBaseType()
}

/**
 * An object in memory which may be identified by the variable that holds it.
 *
 * This may be a local variable, a global variable, or a parameter, etc. However, it cannot be a
 * member of a struct or union, as these do not have storage duration.
 */
class VariableObjectIdentity extends Variable, ObjectIdentityBase {
  VariableObjectIdentity() {
    // Exclude members; member definitions does not allocate storage and thus do not have a storage
    // duration. They are therefore not objects. To get the storage duration of members, use one of
    // the predicates related to sub objects, e.g. `getASubObjectType()`.
    not isMember()
  }

  override StorageDuration getStorageDuration() {
    // 6.2.4.4, objects declared _Thread_local have thread storage duration.
    isThreadLocal() and result.isThread()
    or
    // 6.2.4.3, Non _ThreadLocal objects with internal or external linkage or declared static have
    // static storage duration.
    not isThreadLocal() and
    (hasLinkage() or isStatic()) and
    result.isStatic()
    or
    // 6.2.4.3, Non _ThreadLocal objects no linkage that are not static have automatic storage
    // duration.
    not isThreadLocal() and
    not hasLinkage() and
    not isStatic() and
    result.isAutomatic()
  }

  override Type getType() {
    // Caution here: If we use `Variable.super.getType()` then override resolution is skipped, and
    // it uses the base predicate defined as `none()`. By casting this to `Variable` and calling
    // `getType()`, all overrides (harmlessly, *including this one*...) are considered, which means
    // we defer to the subclasses such as `GlobalVariable` overrides of `getType()`, which is what
    // we want.
    result = this.(Variable).getType()
  }

  /* The storage duration of a variable depends on its linkage. */
  IdentifierLinkage getLinkage() { result = linkageOfVariable(this) }

  predicate hasLinkage() { not getLinkage().isNone() }

  override VariableAccess getAnAccess() { result = Variable.super.getAnAccess() }

  override predicate hasTemporaryLifetime() {
    none() // Objects identified by a variable do not have temporary lifetime.
  }
}

/**
 * A string literal is an object with static storage duration.
 *
 * 6.4.5.6, multibyte character sequences initialize an array of static storage duration.
 */
class LiteralObjectIdentity extends Literal, ObjectIdentityBase {
  override StorageDuration getStorageDuration() { result.isStatic() }

  override Type getType() { result = Literal.super.getType() }

  override Expr getAnAccess() { result = this }

  override predicate hasTemporaryLifetime() {
    none() // String literal objects do not have temporary lifetime.
  }
}

/**
 * An object identifiable as a struct or array literal, which is an lvalue that may have static or
 * automatic storage duration depending on context.
 *
 * 6.5.2.5.5, compound literals outside of a function have static storage duration, while literals
 * inside a function have automatic storage duration.
 */
class AggregateLiteralObjectIdentity extends AggregateLiteral, ObjectIdentityBase {
  override StorageDuration getStorageDuration() {
    if exists(getEnclosingFunction()) then result.isAutomatic() else result.isStatic()
  }

  override Type getType() { result = AggregateLiteral.super.getType() }

  override Expr getAnAccess() { result = this }

  override predicate hasTemporaryLifetime() {
    // Confusing; a struct literal is an lvalue, and therefore does not have temporary lifetime.
    none()
  }
}

/**
 * An object identified by a call to `malloc`.
 *
 * Note: the malloc expression returns an address to this object, not the object itself. Therefore,
 * `getAnAccess()` returns cases where this malloc result is dereferenced, and not the malloc call
 * itself.
 *
 * Note that the predicates for tracking accesses, subobject accesses, and address expresisons may
 * be less reliable as dynamic memory is fundamentally more difficult to track. However, this class
 * attempts to give reasonable results. In the future, this could be improved by integrating with
 * LifetimeProfile.qll or by integrating flow analysis.
 *
 * Additionally, the type of this object is inferred based on its size and use.
 */
class AllocatedObjectIdentity extends AllocationExpr, ObjectIdentityBase {
  AllocatedObjectIdentity() {
    this.(FunctionCall).getTarget().(AllocationFunction).requiresDealloc()
  }

  override StorageDuration getStorageDuration() { result.isAllocated() }

  /** Attempt to infer the type of the allocated memory */
  override Type getType() { result = this.getAllocatedElementType() }

  /** Find dereferences of direct aliases of this pointer result. */
  override Expr getAnAccess() { result.(PointerDereferenceExpr).getOperand() = getAnAlias() }

  /**
   * Find the following subobject accesses, given a pointer alias `x`:
   * - `(*x)`
   * - `(*x).y`
   * - `(*x)[i]`
   * - `x->y`
   * - `x[i]`
   * - `x->y.z`
   * - `x[i].y`
   * - all direct accesses (`foo.x`, `foo[i]`) of the above
   */
  override Expr getASubobjectAccess() {
    result = getASubobjectAccessOf(getAnAccess())
    or
    exists(PointerFieldAccess pfa |
      pfa.getQualifier() = getASubobjectAddressExpr() and
      result = getASubobjectAccessOf(pfa)
    )
    or
    exists(ArrayExpr arrayExpr |
      arrayExpr.getArrayBase() = getASubobjectAddressExpr() and
      result = getASubobjectAccessOf(arrayExpr)
    )
  }

  /**
   * Given a pointer alias `x`, finds `x` itself. Additionally, defers to the default class
   * behavior, which finds address-of (`&`) and array-to-pointer conversions of all subobject
   * accesses. (See `AllocatedObjectIdentity.getASubobjectAccess()`.)
   */
  override Expr getASubobjectAddressExpr() {
    result = getAnAlias()
    or
    result = super.getASubobjectAddressExpr()
  }

  /**
   * Find an obvious direct reference to the result of a `malloc()` function call. This includes
   * the function call itself, but additionally:
   * - For `T* x = malloc(...)`, accesses to variable `x` are likely aliases of the malloc result
   * - For `(expr) = malloc(...)` future lexically identical uses of `expr` are likely aliases of
   *   the malloc result.
   *
   * This is used so that member predicates such as `getAnAccess()`, `getASubobjectAccess()` can
   * find cases such as:
   *
   * ```c
   *   int *x = malloc(sizeof(int));
   *   return *x; // accesses the malloc result
   * ```
   */
  Expr getAnAlias() {
    result = this
    or
    exists(AssignExpr assignExpr |
      assignExpr.getRValue() = this and
      hashCons(result) = hashCons(assignExpr.getLValue())
    )
    or
    exists(Variable v |
      v.getInitializer().getExpr() = this and
      result = v.getAnAccess()
    )
  }

  override predicate hasTemporaryLifetime() {
    none() // Allocated objects do not have "temporary lifetime."
  }
}

/**
 * A struct or union type that contains an array type, used to find objects with temporary
 * lifetime.
 */
private class StructOrUnionTypeWithArrayField extends Struct {
  StructOrUnionTypeWithArrayField() {
    this.getAField().getUnspecifiedType() instanceof ArrayType
    or
    // nested struct or union containing an array type
    this.getAField().getUnspecifiedType().(Struct) instanceof StructOrUnionTypeWithArrayField
  }
}

/**
 * 6.2.4.7, A non-lvalue expression with struct or or union type that has a field member of array
 * type, refers to an object with automatic storage duration (and has temporary lifetime).
 *
 * The spec uses the lanugage "refers to." This is likely intended to mean that the expression
 * `foo().x` does not create a new temporary object, but rather "refers to" the temporary object
 * storing the value of the expression `foo()`.
 *
 * Separate this predicate to avoid non-monotonic recursion (`C() { not exists(C c | ... ) }`).
 */
private class TemporaryObjectIdentityExpr extends Expr {
  TemporaryObjectIdentityExpr() {
    getType() instanceof StructOrUnionTypeWithArrayField and
    not isCLValue(this)
  }
}

/**
 * 6.2.4.7, A non-lvalue expression with struct or or union type that has a field member of array
 * type, is an object with automatic storage duration (and has temporary lifetime).
 */
class TemporaryObjectIdentity extends ObjectIdentityBase instanceof TemporaryObjectIdentityExpr {
  TemporaryObjectIdentity() {
    // See comment in `TemporaryObjectIdentityExpr` for why we check `getASubobjectAccess()` here.
    not exists(TemporaryObjectIdentityExpr parent |
      this = getASubobjectAccessOf(parent) and
      not this = parent
    )
  }

  override StorageDuration getStorageDuration() { result.isAutomatic() }

  override Type getType() { result = this.(Expr).getType() }

  override Expr getAnAccess() { result = this }

  override predicate hasTemporaryLifetime() { any() }
}
