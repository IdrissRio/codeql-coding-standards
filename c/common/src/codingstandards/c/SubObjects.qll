/**
 * A library that expands upon the `Objects.qll` library, to support nested "Objects" such as
 * `x.y.z` or `x[i][j]` within an object `x`.
 *
 * Objects in C are values in memory, that have a type and a storage duration. In the case of
 * array objects and struct objects, the object will contain other objects. The these subobjects
 * will share properties of the root object such as storage duration. This library can be used to,
 * for instance, find all usages of a struct member to ensure that member is initialized before it
 * is used.
 *
 * To use this library, select `SubObject` and find its usages in the AST via `getAnAccess()` (to
 * find usages of the subobject by value) or `getAnAddressOfExpr()` (to find usages of the object
 * by address).
 *
 * Note that a struct or array object may contain a pointer. In this case, the pointer itself is
 * a subobject of the struct or array object, but the object that the pointer points to is not.
 * This is because the pointed-to object does not necessarily have the same storage duration,
 * lifetime, or linkage as the pointer and the object containing the pointer.
 *
 * Note as well that `getAnAccess()` on an array subobject will return all accesses to the array,
 * not just accesses to a particular index. For this reason, `SubObject` exposes the predicate
 * `isPrecise()`. If a subobject is precise, that means all results of `getAnAccess()` will
 * definitely refer to the same object in memory. If it is not precise, the different accesses
 * may refer to the same or different objects in memory. For instance, `x[i].y` and `x[j].y` are
 * the same object if `i` and `j` are the same, but they are different objects if `i` and `j` are
 * different.
 */

import codingstandards.c.Objects

newtype TSubObject =
  TObjectRoot(ObjectIdentity i) or
  TObjectMember(SubObject struct, MemberVariable m) {
    m = struct.getType().(Struct).getAMemberVariable()
  } or
  TObjectIndex(SubObject array) { array.getType() instanceof ArrayType }

class SubObject extends TSubObject {
  string toString() {
    exists(ObjectIdentity i |
      this = TObjectRoot(i) and
      result = i.toString()
    )
    or
    exists(SubObject struct, Variable m |
      this = TObjectMember(struct, m) and
      result = struct.toString() + "." + m.getName()
    )
    or
    exists(SubObject array |
      this = TObjectIndex(array) and
      result = array.toString()
    )
  }

  Type getType() {
    exists(ObjectIdentity i |
      this = TObjectRoot(i) and
      result = i.getType()
    )
    or
    exists(Variable m |
      this = TObjectMember(_, m) and
      result = m.getType()
    )
    or
    exists(SubObject array |
      this = TObjectIndex(array) and
      result = array.getType().(ArrayType).getBaseType()
    )
  }

  /**
   * Holds for object roots and for member accesses on that root, not for array accesses.
   *
   * This is useful for cases where we do not wish to treat `x[y]` and `x[z]` as the same object.
   */
  predicate isPrecise() { not getParent*() = TObjectIndex(_) }

  SubObject getParent() {
    exists(SubObject struct, MemberVariable m |
      this = TObjectMember(struct, m) and
      result = struct
    )
    or
    exists(SubObject array |
      this = TObjectIndex(array) and
      result = array
    )
  }

  Expr getAnAccess() {
    exists(ObjectIdentity i |
      this = TObjectRoot(i) and
      result = i.getAnAccess()
    )
    or
    exists(MemberVariable m |
      this = TObjectMember(_, m) and
      result = m.getAnAccess() and
      // Only consider `DotFieldAccess`es, not `PointerFieldAccess`es, as the latter
      // are not subobjects of the root object:
      result.(DotFieldAccess).getQualifier() = getParent().getAnAccess()
    )
    or
    this = TObjectIndex(_) and
    result.(ArrayExpr).getArrayBase() = getParent().getAnAccess()
  }

  AddressOfExpr getAnAddressOfExpr() { result.getOperand() = this.getAnAccess() }

  /**
   * Get the "root" object identity to which this subobject belongs. For instance, in the
   * expression `x.y.z`, the root object is `x`. This subobject will share properties with the root
   * object such as storage duration, lifetime, and linkage.
   */
  ObjectIdentity getRootIdentity() {
    exists(ObjectIdentity i |
      this = TObjectRoot(i) and
      result = i
    )
    or
    result = getParent().getRootIdentity()
  }
}
