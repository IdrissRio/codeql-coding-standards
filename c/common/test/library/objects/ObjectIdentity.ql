import codingstandards.c.Objects

from ObjectIdentity obj
where obj.getFile().getBaseName() = "objectidentity.c"
select obj, obj.getStorageDuration(), obj.getType()
