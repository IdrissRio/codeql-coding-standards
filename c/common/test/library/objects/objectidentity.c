#include "threads.h"
// Basic static storage duration
int g_statstg1;
extern int g_statstg2;
static int g_statstg3;

// Basic automatic storage duration
void f1(int p_autostg1) { int l_autostg2; }

// Block identifiers with static storage duration
void f2(void) {
  extern int l_statstg1;
  static int l_statstg2;
}

// Thread storage duration
_Thread_local g_thrdstg1;
extern _Thread_local g_thrdstg2;
static _Thread_local g_thrdstg3;
void f3() {
  extern _Thread_local l_statstg3;
  static _Thread_local l_statstg4;
}

// Struct declarations do not allocate storage, and do not have a storage
// duration.
struct s {
  int m;
};

// Enums and enum constants are not variables and have no linkage.
enum e { E1 };

// Various literals:
struct s *g_statstg4 = &(struct s){0};
char *g_statstg5 = "hello";
void f4(void) {
  1;
  "hello";
  (struct s){1};
}