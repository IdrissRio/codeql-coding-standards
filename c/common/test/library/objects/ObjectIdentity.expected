| objectidentity.c:3:5:3:14 | g_statstg1 | static storage duration | file://:0:0:0:0 | int |
| objectidentity.c:4:12:4:21 | g_statstg2 | static storage duration | file://:0:0:0:0 | int |
| objectidentity.c:5:12:5:21 | g_statstg3 | static storage duration | file://:0:0:0:0 | int |
| objectidentity.c:8:13:8:22 | p_autostg1 | automatic storage duration | file://:0:0:0:0 | int |
| objectidentity.c:8:31:8:40 | l_autostg2 | automatic storage duration | file://:0:0:0:0 | int |
| objectidentity.c:12:14:12:23 | l_statstg1 | static storage duration | file://:0:0:0:0 | int |
| objectidentity.c:13:14:13:23 | l_statstg2 | static storage duration | file://:0:0:0:0 | int |
| objectidentity.c:17:15:17:24 | g_thrdstg1 | thread storage duration | file://:0:0:0:0 | int |
| objectidentity.c:18:22:18:31 | g_thrdstg2 | thread storage duration | file://:0:0:0:0 | int |
| objectidentity.c:19:22:19:31 | g_thrdstg3 | thread storage duration | file://:0:0:0:0 | int |
| objectidentity.c:21:24:21:33 | l_statstg3 | thread storage duration | file://:0:0:0:0 | int |
| objectidentity.c:22:24:22:33 | l_statstg4 | thread storage duration | file://:0:0:0:0 | int |
| objectidentity.c:35:11:35:20 | g_statstg4 | static storage duration | file://:0:0:0:0 | s * |
| objectidentity.c:35:35:35:37 | {...} | static storage duration | objectidentity.c:27:8:27:8 | s |
| objectidentity.c:35:36:35:36 | 0 | static storage duration | file://:0:0:0:0 | int |
| objectidentity.c:36:7:36:16 | g_statstg5 | static storage duration | file://:0:0:0:0 | char * |
| objectidentity.c:36:20:36:26 | hello | static storage duration | file://:0:0:0:0 | char[6] |
| objectidentity.c:38:3:38:3 | 1 | static storage duration | file://:0:0:0:0 | int |
| objectidentity.c:39:3:39:9 | hello | static storage duration | file://:0:0:0:0 | char[6] |
| objectidentity.c:40:3:40:15 | {...} | automatic storage duration | objectidentity.c:27:8:27:8 | s |
| objectidentity.c:40:14:40:14 | 1 | static storage duration | file://:0:0:0:0 | int |
